test/Test/API.purs

```diff
@@ -0,0 +1,197 @@
+module Test.API where
+
+import Prelude
+
+import Control.Alternative (class Alternative)
+import Control.Lazy (class Lazy)
+import Control.Monad.Rec.Class (class MonadRec)
+import Data.Foldable (class Foldable)
+import Data.Maybe (Maybe)
+import Data.Tuple (Tuple)
+import Data.Unfoldable (class Unfoldable)
+
+type Common c =
+  { makeCollection :: forall f a. Foldable f => f a -> c a
+
+  , concat :: forall a. c (c a) -> c a
+  , concatMap :: forall a. forall b. (a -> c b) -> c a -> c b
+  , cons :: forall a. a -> c a -> c a
+  , elemIndex :: forall a. Eq a => a -> c a -> Maybe Int
+  , elemLastIndex :: forall a. Eq a => a -> c a -> Maybe Int
+  , findIndex :: forall a. (a -> Boolean) -> c a -> Maybe Int
+  , findLastIndex :: forall a. (a -> Boolean) -> c a -> Maybe Int
+  , foldM :: forall m a b. Monad m => (b -> a -> m b) -> b -> c a -> m b
+  , index :: forall a. c a -> Int -> Maybe a
+  , intersect :: forall a. Eq a => c a -> c a -> c a
+  , intersectBy :: forall a. (a -> a -> Boolean) -> c a -> c a -> c a
+  , length :: forall a. c a -> Int
+  , nubEq :: forall a. Eq a => c a -> c a
+  , nubByEq :: forall a. (a -> a -> Boolean) -> c a -> c a
+  , range :: Int -> Int -> c Int
+  , reverse :: c ~> c
+  , singleton :: forall a. a -> c a
+  , snoc :: forall a. c a -> a -> c a
+  , toUnfoldable :: forall f a. Unfoldable f => c a -> f a
+  , union :: forall a. Eq a => c a -> c a -> c a
+  , unionBy :: forall a. (a -> a -> Boolean) -> c a -> c a -> c a
+  , unzip :: forall a b. c (Tuple a b) -> Tuple (c a) (c b)
+  , zip :: forall a b. c a -> c b -> c (Tuple a b)
+  , zipWith :: forall a b d. (a -> b -> d) -> c a -> c b -> c d
+  , zipWithA :: forall a b d m. Applicative m => (a -> b -> m d) -> c a -> c b -> m (c d)
+
+  , appendFoldable :: forall t a. Foldable t => c a -> t a -> c a
+  , insert :: forall a. Ord a => a -> c a -> c a
+  , insertBy :: forall a. (a -> a -> Ordering) -> a -> c a -> c a
+  , nub :: forall a. Ord a => c a -> c a
+  , nubBy :: forall a. (a -> a -> Ordering) -> c a -> c a
+  -- , replicate :: forall a. Int -> a -> c a
+  , some :: forall f a. Alternative f => Lazy (f (c a)) => f a -> f (c a)
+  , someRec :: forall f a. MonadRec f => Alternative f => f a -> f (c a)
+  , sort :: forall a. Ord a => c a -> c a
+  , sortBy :: forall a. (a -> a -> Ordering) -> c a -> c a
+  , transpose :: forall a. c (c a) -> c (c a)
+  }
+
+type CommonDiffEmptiability c cInverse canEmpty nonEmpty cPattern =
+  { makeCollection :: forall f a. Foldable f => f a -> c a
```

Comment:
To reduce verbosity and help show similarities (though I don't think this would help readability), should items like `makeCollection` be stored as a `Row` that gets added into each differing item via row polymorphism?

Reply:
That could be a nice adjustment, but not sure the deduplication would be enough to justify the readability tradeoff and Row + Record overhead. I might be misimagining what changes are involved with this proposal though, so would be interested in seeing it implemented.

Another option could be to do something like this:
https://github.com/milesfrain/purescript-lists/pull/4/files

Reply:
Yeah... I'm not sure it's worth it either due to the readability tradeoff.